batch/batch/worker/worker.py

```diff
@@ -425,138 +490,271 @@ async def batch_worker_access_token(self):
-                    self.job.batch_id,
-                    self.job.job_id,
-                    self.job.attempt_id,
-                    self.name,
-                    await self.get_container_log(),
-                )
+    async def write_container_config(self):
+        os.makedirs(self.config_path)
+        with open(f'{self.config_path}/config.json', 'w') as f:
+            f.write(json.dumps(await self.container_config()))
 
-            with self.step('deleting'):
-                await self.delete_container()
+    # https://github.com/opencontainers/runtime-spec/blob/master/config.md
+    async def container_config(self):
+        uid, gid = await self._get_in_container_user()
+        weight = worker_fraction_in_1024ths(self.spec['cpu'])
+        default_docker_capabilities = [
+            'CAP_CHOWN',
+            'CAP_DAC_OVERRIDE',
+            'CAP_FSETID',
+            'CAP_FOWNER',
+            'CAP_MKNOD',
+            'CAP_NET_RAW',
+            'CAP_SETGID',
+            'CAP_SETUID',
+            'CAP_SETFCAP',
+            'CAP_SETPCAP',
+            'CAP_NET_BIND_SERVICE',
+            'CAP_SYS_CHROOT',
+            'CAP_KILL',
+            'CAP_AUDIT_WRITE',
+        ]
+        config = {
+            'ociVersion': '1.0.1',
+            'root': {
+                'path': '.',
+                'readonly': False,
+            },
+            'hostname': self.netns.hostname,
+            'mounts': self._mounts(),
+            'process': {
+                'user': {  # uid/gid *inside the container*
+                    'uid': int(uid),
+                    'gid': int(gid),
+                },
+                'args': self.spec['command'],
+                'env': self._env(),
+                'cwd': '/',
+                'capabilities': {
+                    'bounding': default_docker_capabilities,
+                    'effective': default_docker_capabilities,
+                    'inheritable': default_docker_capabilities,
+                    'permitted': default_docker_capabilities,
+                },
+            },
+            'linux': {
+                'namespaces': [
+                    {'type': 'pid'},
+                    {
+                        'type': 'network',
+                        'path': f'/var/run/netns/{self.netns.network_name}',
+                    },
+                    {'type': 'mount'},
+                    {'type': 'ipc'},
+                    {'type': 'uts'},
+                    {'type': 'cgroup'},
+                ],
+                'uidMappings': [],  # Can control host -> container uid/gid mappings here
+                'gidMappings': [],
+                'resources': {
+                    'cpu': {'shares': weight},
+                    'memory': {
+                        'limit': self.spec['memory'],
+                        'reservation': self.spec['memory'],
+                    },
+                    # 'blockIO': {'weight': min(weight, 1000)}, FIXME blkio.weight not supported
```

Comment:
OK, so, let's pin down exactly the issues here.

`blkio.weight` was deprecated and then removed in Liinux 5.0. The cgroups man page (what a beauty of
a man page) notes this:

> Although cgroups v2 is intended as a replacement for cgroups v1, the older system continues to
> exist (and for compatibility reasons is unlikely to be removed).  Currently, cgroups v2 implements
> only a subset of the controllers available in cgroups v1.  The two systems are implemented so that
> both v1 controllers and v2 controllers can be mounted on the same system.  Thus, for example, it
> is possible to use those controllers that are supported under version 2, while also using version
> 1 controllers where version 2 does not yet support those controllers.  The only restriction here
> is that a controller can't be simultaneously employed in both a cgroups v1 hierarchy and in the
> cgroups v2 hierarchy.

[cgroups(7)](https://man7.org/linux/man-pages/man7/cgroups.7.html)

Supposedly, the new `io` cgroup still supports `weight`:

> blkio.weight was removed in kernel 5.0: torvalds/linux@f382fb0
>
> cgroup v2 still has io.weight which can be converted from blkio.weight

[AkihiroSuda](https://github.com/docker/cli/pull/2908#issuecomment-755867349)

It appears that batch workers might support cgroup2:

```
dking@batch-worker-default-standard-n2-d7sea:~$ mount | grep cgroup
...
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
...
```

Clearly some more work is needed to understand this, but I'll leave this comment here as
documentation about what we need to do next.